doc/developer-guide/io-framework.md

```diff
@@ -0,0 +1,294 @@
+  approach. This means that all I/O will be done in background without
+  blocking the worker threads.
+
+_io_uring_ is only present on latest linux kernels and it's dynamically
+detected and used if available. Otherwise it silently fails back to the
+synchronous implementation in a transparent way for the rest of the code
+that uses this framework (once implemented, it will use the threaded I/O
+instead of the legacy mode when io_uring is not supported).
+
+The implementation is done with the io_uring API in mind. This means that
+io_uring fits very well in the I/O framework API, and the other modes
+are adjusted to follow the same semantics.
+
+## How to use it
+
+In this section a general overview of the operation will be provided,
+focused on the io_uring-based implementation. For differences when
+io_uring is not present, check section [Fallback mode](#fallback-mode)
+
+### Initialization
+
+The framework is initialized using `gf_io_run()`.
+
+```c
+typedef int32_t (*gf_io_async_t)(gf_io_op_t *op);
+
+typedef struct {
+    gf_io_async_t setup;
+    gf_io_async_t cleanup;
+} gf_io_handlers_t;
+
+int32_t gf_io_run(gf_io_handlers_t handlers, void *data);
+```
+
+The handlers structure contains two async functions, one that is called
+just after having initialized the I/O infrastructure, and another one that
+is called after stopping everything else. The 'data' argument is an extra
+argument that will be passed to each function.
+
+The returned value can be a negative error code if there has been any
+problem while initializing the system, or 0 if everything worked fine. In
+this case, the function only returns when the program is terminating.
+
+### Termination
+
+When it's determined that the process must be terminated, a call to
+`gf_io_shutdown()` must be done.
+
+```c
+void gf_io_shutdown(void);
+```
+
+This function initiates a shutdown procedure, but returns immediately.
+Once the shutdown is completed, `gf_io_run()` will return. It can be
+called from anywhere.
+
+When shutdown is initiated, all I/O should have been stopped. If there
+is active I/O during the shutdown, they can complete, fail or be cancelled,
+depending on what state the request was. To ensure consistent behavior, try
+to always stop I/O before terminating the I/O framework.
+
+### Normal operation
+
+After everything is ready, the normal operation of the I/O framework is
+very simple:
+
+1. A worker picks one completion event from the kernel.
+
+2. The callback associated to the completion event is executed.
+
+   2.1. The callback can prepare new I/O requests using one of the
+        `gf_io_*` I/O functions available for I/O operations.
+
+   2.2. Requests can be sent one by one or submitted in a batch. In all
+        cases they are added to the io_uring SQ ring.
+
+   2.3. The callback may (but it's not required to) explicitly flush all
+        queued requests to the kernel using `gf_io_flush()`.
```

Comment:
Yes. I initially though about providing a public interface to the `flush` operation, but since I haven't needed it yet, it's not implemented. Anyway it would be a simple wrapped that would call engine's `flush` function.

Reply:
After a second thought, I've removed this part. The way the thread pool is implemented guarantees that all threads are continuously processing requests. Providing a flush function that can be called anywhere means that the code after flush may need to wait for an unbounded amount of time, which could be bad. Flush is automatically done frequently enough by all threads, so it shouldn't be necessary. Only requests sent from outside the thread pool may need an explicit flush, but this will be handled internally by specific functions.